File: LeetCodeAlgos/Java/KthLargestElementInStream.java
import java.util.PriorityQueue;

public class KthLargestElementInStream {
  public static void main(String[] args) {
    KthLargestElementInStream stream = new KthLargestElementInStream(3, new int[]{4, 5, 8, 2});
    testLog(4, stream.add(3));
    testLog(5, stream.add(5));
    testLog(5, stream.add(10));
    testLog(8, stream.add(9));
    testLog(8, stream.add(4));

    stream = new KthLargestElementInStream(4, new int[]{7, 7, 7, 7, 8, 3});
    testLog(7, stream.add(2));
    testLog(7, stream.add(10));
    testLog(7, stream.add(9));
    testLog(8, stream.add(9));
  }
  public static void testLog(int expected, int actual){
    System.out.printf("%d == %d (%b)\n", expected, actual, expected == actual);
  }



  PriorityQueue<Integer> queue;
  int k;
  public KthLargestElementInStream(int k, int[] nums) {
    this.k = k;
    this.queue = new PriorityQueue<>(k, Integer::compareTo);
    for (int num : nums) {
      this.add(num);
    }
  }

  public int add(int val) {
    if (queue.size() >= this.k){
      if (queue.peek() > val) {
        return queue.peek();
      }
      queue.poll();
    }
    queue.offer(val);
    return queue.peek();
  }
}

File: LeetCodeAlgos/Rust/kth-largest-element-in-stream.rs
// You are part of a university admissions office and need to keep track of the kth highest test score from applicants in real-time. This helps to determine cut-off marks for interviews and admissions dynamically as new applicants submit their scores.

// You are tasked to implement a class which, for a given integer k, maintains a stream of test scores and continuously returns the kth highest test score after a new score has been submitted. More specifically, we are looking for the kth highest score in the sorted list of all scores.

use std::collections::BinaryHeap;
use std::cmp::Reverse;
struct KthLargest {
  k: i32,
  heap: BinaryHeap<Reverse<i32>>
}


/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl KthLargest {
    fn new(k: i32, nums: Vec<i32>) -> Self {
      let mut new = Self {
        k,
        heap: BinaryHeap::<Reverse<i32>>::with_capacity(k as usize)
      };
      for num in nums {
        new.add(num);
      }
      new
    }

    fn add(&mut self, val: i32) -> i32 {
      match self.heap.len() == self.heap.capacity() {
      true => {
        if &Reverse(val) > self.heap.peek().unwrap() {
          self.heap.pop();
          self.heap.push(Reverse(val));
        }
      }
      false => { self.heap.push(Reverse(val)); }
    }
    self.heap.peek().unwrap().0
  }
}